checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,873 @@
+   * whose must-call obligation is satisfied within the current method. Those fields are the ones in
+   * variable {@link #disposedFields}.
+   */
+  private void addEnsuresCalledMethods() {
+    // The keys are the must-call method names, and the values are the set of fields on which those
+    // methods should be called. This map is used to create @EnsuresCalledMethods annotation for
+    // fields that share the same must-call obligation.
+    Map<String, Set<String>> methodToFields = new LinkedHashMap<>();
+    for (VariableElement disposedField : disposedFields) {
+      List<String> mustCallValues = resourceLeakAtf.getMustCallValues(disposedField);
+      assert !mustCallValues.isEmpty()
+          : "Must-call obligation of owning field " + disposedField + " is empty.";
+      // Currently, this code assumes that the must-call set has only one element.
+      assert mustCallValues.size() == 1
+          : "The must-call set of " + disposedField + "should be a singleton: " + mustCallValues;
+      String mustCallValue = mustCallValues.get(0);
+      String fieldName = "this." + disposedField.getSimpleName().toString();
+
+      methodToFields.computeIfAbsent(mustCallValue, k -> new HashSet<>()).add(fieldName);
+    }
+
+    for (String mustCallValue : methodToFields.keySet()) {
+      Set<String> fields = methodToFields.get(mustCallValue);
+      AnnotationMirror am =
+          createEnsuresCalledMethods(
+              fields.toArray(new String[fields.size()]), new String[] {mustCallValue});
+      WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+      wpi.addMethodDeclarationAnnotation(methodElt, am);
+    }
+  }
+
+  /**
+   * Possibly adds an InheritableMustCall annotation on the enclosing class. If the class already
+   * has a non-empty MustCall type (that is inherited from one of its superclasses), this method
+   * does nothing, in order to avoid infinite iteration. Otherwise, if the current method is not
+   * private and satisfies the must-call obligations of all the owning fields, it adds (or updates)
+   * an InheritableMustCall annotation to the enclosing class.
+   */
+  private void addOrUpdateMustCall() {
+    if (classElt == null) {
+      return;
+    }
+
+    WholeProgramInference wpi = resourceLeakAtf.getWholeProgramInference();
+    List<String> currentMustCallValues = resourceLeakAtf.getMustCallValues(classElt);
+    if (!currentMustCallValues.isEmpty()) {
+      // The class already has a MustCall annotation.
+
+      // If it is inherited from a superclass, do nothing.
+      if (classElt.getSuperclass() != null) {
+        TypeMirror superType = classElt.getSuperclass();
+        TypeElement superClassElt = TypesUtils.getTypeElement(superType);
+        if (superClassElt != null && !resourceLeakAtf.getMustCallValues(superClassElt).isEmpty()) {
+          return;
+        }
+      }
+      // Add the existing @MustCall annotation to guarantee the termination of inference. This is
```

Comment:
Add to what?  If a `@MustCall` annotation already exists, what is the point of re-adding it?  It seems that such an operation would have no effect, so "add" must mean something else in this context.  What is the effect?

Reply:
If the enclosing class already has a non-empty @MustCall type, either added by programmers or inferred in previous iterations (not-inherited), we avoid recomputing it in the current round to prevent potential inconsistencies when there are multiple candidates for the @MustCall annotation. So,  if not re-added, it won't appear in the inferred result for this iteration.